Release pull-request-details-action Github Action

elliot-nelson opened the pull request:
### SUMMARY

Initial implementation, unit tests, and publishing steps for the action.

### DETAILS

#### Implementation

The action implementation is straightforward and was taken from the existing inline implementations in our internal repos.

#### Unit testing

For unit tests, I'm following GitHub's examples and using `process.env` as the API for passing inputs into the action, and using standard function spies to verify the behavior inside the action.

#### Publishing

The publishing script is designed to meet the goals we agreed on in a previous OSSWG discussion -- users of the action should be able to use a "GitHub-style" major version (`pull-request-details-action@v1`), a specific tagged version (`pull-request-details-action@v1.0.3`), or a commit hash (`pull-request-details-action@aaaa....`), depending on their comfort level with potential changes. However, we don't want our main branch muddied up with git diffs in the `dist/` (bundled code) folder, and we definitely don't want the alternative (checking in node_modules).

This script accomplishes this by:

 - From `main`, we branch off a snapshot branch (`snapshot/v1.0.3`). 
 - We comment out the `dist/` folder in `.gitignore`, run the `ncc` bundler, and check in the results.
 - This commit is tagged with the version tag (`v1.0.3`) and pushed to GitHub.
 - We then reset and force-push the associated major version branch (`v1`), to point it at the latest version of `v1`.

Note that we never bundle in `main`, so a caveat is that the use of `pull-request-details-action@main` is not supported.  I think that's an acceptable trade-off for having a clean main branch, but we probably should document that in the README.

### CHECKLIST
- [x] Documentation updated (if needed)
- [x] Unit tests exist to cover the code you are changing and validated
- [ ] Integration tests exist to cover the code you are changing and validated

elliot-nelson committed:
Initial markdown files

elliot-nelson committed:
Action contents and publishing script

elliot-nelson committed:
Initial unit tests

elliot-nelson committed:
Improve package script names

elliot-nelson committed:
Publishing is a no-op if version is published

octogonz commented on lib/index.js:
```diff
@@ -0,0 +1,34 @@
+    core.setOutput('head-ref',   data.head.ref);
+    core.setOutput('head-sha',   data.head.sha);
+    core.setOutput('head-owner', data.head.repo.owner.login);
+    core.setOutput('head-repo',  data.head.repo.name);
```
Are all these fields guaranteed to exist?  It's hard to tell without TypeScript

elliot-nelson replied:
Yes (if we find a PR, the `data` returned is consistent.)

octogonz commented on README.md:
```diff
@@ -0,0 +1,30 @@
+# pull-request-details-action
+
+Retrieve the branch details of a specific pull request.
```
For somebody who's new to GitHub Actions, where do the retrieved results go?

Might be useful to explain:
- That GitHub actions allow steps to have an `id` such that `steps.<id>.outputs.<variable>` can be accessed using the `${{ }}`.  Maybe briefly summarize the idea and link to GitHub's docs for details.
- What are the list of outputs produced by `pull-request-details-action`?
- For each output field, give a short description and an sample value (to indicate the syntax)

octogonz replied:
(Some of this info is in `action.yaml` but that is not easily discoverable in the decs.)

octogonz commented on README.md:
```diff
@@ -0,0 +1,30 @@
+
+Retrieve the branch details of a specific pull request.
+
+This action is not useful for workflows triggered by pull request events, where information about the head and base branches of the pull request is already available. However, some events (such as issue comments) do not include the full SHA information for both the head and the base branch, and in those cases this action can be helpful.
```
This note would be more understandable with concrete examples.  *"Suppose we're making a GitHub Action that does ___.  In this case we can determine ___ easily by doing ___.  But suppose instead that your GitHub Action does ___.  In this case ___ is not available because ___.  That's where `pull-request-details-action` helps out!"*

octogonz replied:
In other words, think from the perspective of a person who would benefit from `pull-request-details-action` if only they new about your project. What would they be searching for on Google?  Probably more specific keywords than "Retrieve the branch details of a specific pull request"

octogonz commented on lib/index.js:
```diff
@@ -0,0 +1,34 @@
+module.exports = { main };
+
+/* istanbul ignore if */
+if (require.main === module) {
```
What does this do?  A comment would be helpful

octogonz commented on package.json:
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "pull-request-details-action",
+  "version": "1.0.5",
```
Why are we starting with `1.0.5`?

elliot-nelson replied:
While testing the publishing script I ate several tags.  (If we really wanted to we could overwrite them to release the first version as `1.0.0`, but I think it's actually more realistic to have a couple of patches... either way, the "normal GitHub usage" will be `@v1`, the major-branch approach.)

octogonz commented on package.json:
```diff
@@ -0,0 +1,29 @@
+  "scripts": {
+    "test": "jest",
+    "bundle": "ncc build index.js --license licenses.txt",
+    "publish": "scripts/publish.js"
```
Is this meant to completely replace the conventional `npm publish` operation for making an NPM package?  If so that seems counterintuitive.  Is it normal for GitHub Actions?  Conventionally a **package.json** that is not meant to go to the NPM registry should have `"private": true` to prevent accidental publishing.

elliot-nelson replied:
This is a good question.  I called the script `publish` just because "publishing" the action is familiar terminology, although it has little similarity to the built-in `npm publish`.  I'll add the private tag for sure, to at least avoid someone thinking this line _overrides_ `npm publish` behavior.

octogonz replied:
Maybe rename it to `npm run release` or `npm run publish-github-action`?

elliot-nelson committed:
PR feedback

elliot-nelson commented:
Latest changes:

 - Expanded README with new sections describing the Inputs & Outputs, clarified intro, added Changelog & Contributions sections.
 - Set private tag in package.json
 - Added a comment to the index.js `require.main` boilerplate.

octogonz reviewed (approved):
Approved with one suggestion

elliot-nelson committed:
Add azure-pipelines config, rename publish task

elliot-nelson committed:
Fix for coverage details

elliot-nelson committed:
Set coverageReporters in jest config

elliot-nelson committed:
Remove reportDirectory option
